Support multiple urls

rintcius opened the pull request:
[ch6217]

rintcius commented on core/src/main/scala/quasar/datasource/url/Http4sUrlDatasourceModule.scala:
```diff
@@ -195,12 +209,21 @@ class Http4sUrlDatasourceModule(clientBuilder: Http4sClientBuilder)
+      client: Client[F],
+      requests: IList[UriRequest])
+      : F[Condition[InitializationError[Unit]]] =
+    requests.foldLeft(Condition.normal[InitializationError[Unit]]().point[F]) { case (acc, req) =>
```
may be worth to do this in parallel

rintcius commented on core/src/main/scala/quasar/datasource/url/Http4sUrlDatasourceModule.scala:
```diff
@@ -116,20 +99,19 @@ class Http4sUrlDatasourceModule(clientBuilder: Http4sClientBuilder)
+          urlConfig.requests.zipWithIndex.foldLeft(IMap.empty[ResourcePath, F[InternalResult[F]]])
+            { case (acc, (req, i)) =>
+              val n: ResourceName =
+                name(req.url).getOrElse(defaultName(urlConfig.resourceType, i))
```
The resource paths that are generated is the path of the url, and if that's empty the hostname. If both are empty then a default name of the form `url_${i}.(ld)json` is used. If there are duplicates then the names will be prefixed with a `_`. So this will always result in resources directly under root.

We can consider other templates though. E.g. 2 level resource paths of the form `hostname:port / urlPath` (and take care of non-existing hosts, ports and paths). This will group by `hostname:port` which may be useful in some scenarios.

Adding more levels could work too, but there's a trade-off because adding levels also results in more clicks in the UI.

wemrysi replied:
How about just always making it `hostname/path` in order to avoid path conflicts?

rintcius replied:
Happy to change to `hostname/path` but I think we still have to deal with path conflicts.
E.g. suppose there's just a difference in `query` part of URL or even a different header.

wemrysi replied:
Yeah, almost makes me wonder if we should just store a name for each resource and have the frontend just prepopulate the field using the heuristics you've defined.

rintcius commented:
Depends on https://github.com/slamdata/quasar/pull/4116

wemrysi commented on core/src/main/scala/quasar/datasource/url/UrlDatasourceConfig.scala:
```diff
@@ -16,35 +16,55 @@
     url: Uri,
+    headers: IMap[CaseInsensitiveString, String])
+
+final case class UrlDatasourceConfig(
```
Hm, I wonder if we should have `resourceType` and `compressionScheme` be per-URL as well.

rintcius replied:
Yeah, I didn't like the duplication that would involve in entering the config, My thought was that a customer would typically use 1 combination, and if they don't then it wouldn't be too bad to create a new datasource.

beckyconning replied:
baby steps. if a prospective customer asks for those per url we’ll add that capability.

wemrysi replied:
We could do both, though. i.e. it seems like an arbitrary distinction on the implementation side that compression and format aren't per url while headers are.

However, UX is also important, but we can present a single option for the user and just set the same value for all options.

rintcius replied:
I already changed the server model to just allow multiple URLs.

I think being more flexible in the server than in UI won't work that well because it's now possible to edit datasources too (so then we can't present all valid states properly in UI)

wemrysi commented on core/src/main/scala/quasar/datasource/url/UrlDatasourceConfig.scala:
```diff
@@ -16,35 +16,55 @@
   val DefaultMaxRedirects: Int = 5
 
-  implicit val decodeJson: DecodeJson[UrlDatasourceConfig] =
+  implicit val decodeUrl: DecodeJson[Uri] = DecodeJson.StringDecodeJson.flatMap(s =>
```
`decodeUrl` and `decodeHeaders` should probably be `private`.

rintcius commented:
@wemrysi A couple of changes in this PR after your review

- changed format to allow only multiple urls, not headers
- bumped the version in DatasourceType
- changed the way urls are mapped to resource paths (with a different way to ensure the paths are unique)

wemrysi commented:
We really need to work out how to best deal with config versioning since we've needed it in two places already.

cryogenian commented:
:+1:

rintcius commented:
@wemrysi ready for another look. Had to make sure that single URL still work in the old way to avoid having to migrate in frontend.

djspiewak commented:
@rintcius Does this still need to be **breaking**?

wemrysi commented:
@rintcius I think @djspiewak is correct that this could now be `feature`. Also, LGTM, feel free to merge whenever it makes sense to do so.

rintcius commented:
@wemrysi @djspiewak I think it's still breaking because this PR makes sure `url` is auto-upgraded to `urls`. The FE only works with `urls`, not `url` anymore.

djspiewak commented:
@rintcius Ah that's fair.
